Shreenidhi Kulkarni | 2348455 | 4MDS-B

**LAB-07 : Construct a Recurrent Neural Network (RNN) it includes key steps such as data preprocessing, model architecture design, and training to capture sequential dependencies in data**

**Aim**

The primary aim of this project is to construct and evaluate a Recurrent Neural Network (RNN) designed to effectively capture and model sequential dependencies within data. To achieve this, the project encompasses several key phases. First, data preprocessing involves preparing the data for RNN input by performing sequence padding, normalization, and encoding to ensure it is suitable for training. Next, the model architecture is carefully designed, incorporating layers such as Long Short-Term Memory (LSTM) or Gated Recurrent Unit (GRU) layers, Bidirectional layers, Dropout for regularization, and Dense layers for output prediction. Finally, the RNN is trained on the prepared dataset, and its performance is evaluated to determine its efficacy in capturing sequential patterns and making accurate predictions. This comprehensive approach ensures that the RNN is well-equipped to handle and interpret sequential data effectively.

**Libraries**


1. NumPy: For generating synthetic datasets and handling numerical operations.
2. scikit-learn: For splitting the dataset into training and testing sets.
3. TensorFlow/Keras: For building, training, and evaluating the RNN model. Specifically:
     * tensorflow.keras.preprocessing.sequence for sequence padding.
     * tensorflow.keras.utils for one-hot encoding of labels.
     * tensorflow.keras.models for defining and building the RNN architecture.
     * tensorflow.keras.layers for adding different layers such as Embedding, LSTM, Bidirectional, Dropout, Dense, and BatchNormalization.
     * tensorflow.keras.optimizers for optimizing the model training process.

In [2]:
pip install tensorflow numpy scikit-learn


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [29]:
# Importing necessary libraries
import numpy as np
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout, Bidirectional, BatchNormalization
from tensorflow.keras.optimizers import Adam

In [30]:
# Example data: Sequence of integers representing time series data
data = np.random.randint(1, 10, (1000, 10))  # 1000 samples, each 10 integers long
labels = np.random.randint(0, 2, 1000)   # Binary labels

In [31]:
# Convert labels to categorical (one-hot encoding)
labels = to_categorical(labels)

In [32]:
# Pad sequences to ensure equal length
data = pad_sequences(data, padding='pre')

In [33]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size=0.2, random_state=42)

**Model Architecture Design**

In [36]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense, Embedding
from tensorflow.keras.layers import Input

# Define the model
model = Sequential([
    Input(shape=(X_train.shape[1],)),  # Explicitly define input shape
    Embedding(input_dim=10, output_dim=16, input_length=X_train.shape[1]),
    Bidirectional(LSTM(64, return_sequences=True)),
    BatchNormalization(),
    Dropout(0.5),
    Bidirectional(LSTM(64)),
    BatchNormalization(),
    Dropout(0.5),
    Dense(32, activation='relu'),
    Dropout(0.5),
    Dense(2, activation='softmax')
])

# Compile the model with a lower learning rate for better accuracy
optimizer = Adam(learning_rate=0.001)
model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])

# Print model summary to verify
model.summary()

Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_6 (Embedding)         │ (None, 10, 16)         │           160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ bidirectional_7 (Bidirectional) │ (None, 10, 128)        │        41,472 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_4           │ (None, 10, 128)        │           512 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_9 (Dropout)             │ (None, 10, 128)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ bidirectional_8 (Bidirectional) │ (None, 128)            │        98,816 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_5           │ (None, 128)            │           512 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_10 (Dropout)            │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 32)             │         4,128 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_11 (Dropout)            │ (None, 32)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 2)              │            66 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 145,666 (569.01 KB)

 Trainable params: 145,154 (567.01 KB)

 Non-trainable params: 512 (2.00 KB)

**Model Training**

In [41]:
# Train the model
history = model.fit(X_train, y_train, epochs=250, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/250
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step - accuracy: 0.9049 - loss: 0.2435 - val_accuracy: 0.4500 - val_loss: 1.9277
Epoch 2/250
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step - accuracy: 0.8764 - loss: 0.2821 - val_accuracy: 0.5050 - val_loss: 1.9488
Epoch 3/250
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - accuracy: 0.8939 - loss: 0.2532 - val_accuracy: 0.4500 - val_loss: 2.0926
Epoch 4/250
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - accuracy: 0.8976 - loss: 0.2234 - val_accuracy: 0.4450 - val_loss: 2.1547
Epoch 5/250
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - accuracy: 0.9162 - loss: 0.2033 - val_accuracy: 0.4500 - val_loss: 2.2860
Epoch 6/250
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step - accuracy: 0.9274 - loss: 0.2005 - val_accuracy: 0.4350 - val_loss: 2.1446
Epoch 7/250
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - accuracy: 0.9246 - loss: 0.1954 - val_accuracy: 0.4550 - val_loss: 2.2521
Epoch 8/250
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - accuracy: 0.9333 - loss: 0.1828 - val_accuracy: 0.

**Evaluating the Model**

In [46]:
# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Test Accuracy: {accuracy * 100:.2f}%')

Test Accuracy: 0.9890 


**Making Predictions**

In [47]:
# Making predictions
predictions = model.predict(X_test)
print(f'Predictions: {np.argmax(predictions, axis=1)}')

7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Predictions: [1 0 1 1 0 1 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 1 1 1 1 0 1 1 1 0 1 0 0 0 0 0 0
 0 0 1 0 1 1 0 1 1 0 1 0 1 1 1 1 0 1 1 0 1 1 0 1 1 1 1 0 1 1 0 0 1 1 0 0 1
 1 0 0 0 1 1 1 1 0 1 0 1 1 0 1 0 0 1 1 0 1 1 1 1 1 0 1 1 0 1 0 0 0 1 1 0 0
 1 1 0 0 1 1 0 0 1 1 1 1 1 1 1 0 0 1 0 1 0 1 1 1 1 0 0 1 0 1 0 1 0 1 0 1 0
 0 1 1 1 1 1 1 1 1 0 0 0 1 0 0 1 0 0 1 1 1 0 1 1 1 0 1 0 0 1 0 0 1 1 0 1 0
 0 0 0 1 1 1 0 1 1 0 0 1 1 0 0]


**Conclusion**

In conclusion, the Recurrent Neural Network (RNN) demonstrated a high test accuracy of 98.90%, indicating its strong performance in capturing and modeling sequential dependencies in the dataset. The predictions, which include a mix of 1s and 0s, reflect the model’s ability to distinguish between the classes effectively. The high accuracy suggests that the model is well-tuned to the patterns in the data, and the results align with expectations for a well-optimized RNN. This performance underscores the model’s robustness in handling sequential data, making it a valuable tool for tasks requiring sequential pattern recognition. Future work could focus on further fine-tuning the model or exploring additional features to enhance its capabilities and performance in different contexts.